com.unity.mobile.notifications/Runtime/iOS/Plugins/UnityNotificationData.m

```diff
@@ -0,0 +1,159 @@
+    {
+        notificationData->triggerType = CALENDAR_TRIGGER;
+
+        UNCalendarNotificationTrigger* calendarTrigger = (UNCalendarNotificationTrigger*)request.trigger;
+        NSDateComponents* date = calendarTrigger.dateComponents;
+
+        notificationData->calendarTriggerYear = (int)date.year;
+        notificationData->calendarTriggerMonth = (int)date.month;
+        notificationData->calendarTriggerDay = (int)date.day;
+        notificationData->calendarTriggerHour = (int)date.hour;
+        notificationData->calendarTriggerMinute = (int)date.minute;
+        notificationData->calendarTriggerSecond = (int)date.second;
+    }
+    else if ([request.trigger isKindOfClass: [UNLocationNotificationTrigger class]])
+    {
+#if defined(UNITY_USES_LOCATION) && UNITY_USES_LOCATION
+        notificationData->triggerType = LOCATION_TRIGGER;
+
+        UNLocationNotificationTrigger* locationTrigger = (UNLocationNotificationTrigger*)request.trigger;
+        CLCircularRegion *region = (CLCircularRegion*)locationTrigger.region;
+
+        notificationData->locationTriggerCenterX = region.center.latitude;
+        notificationData->locationTriggerCenterY = region.center.longitude;
+        notificationData->locationTriggerRadius = region.radius;
+        notificationData->locationTriggerNotifyOnExit = region.notifyOnEntry;
+        notificationData->locationTriggerNotifyOnEntry = region.notifyOnExit;
+#endif
+    }
+    else if ([request.trigger isKindOfClass: [UNPushNotificationTrigger class]])
+    {
+        notificationData->triggerType = PUSH_TRIGGER;
+    }
+
+    if ([NSJSONSerialization isValidJSONObject: [request.content.userInfo objectForKey: @"data"]])
+    {
+        NSError *error;
+        NSData *data = [NSJSONSerialization dataWithJSONObject: [request.content.userInfo objectForKey: @"data"]
+                        options: NSJSONWritingPrettyPrinted
+                        error: &error];
+        if (!data)
+        {
+            NSLog(@"Failed parsing notification userInfo[\"data\"]: %@", error);
+        }
+        else
+        {
+            notificationData->data = (char*)[[[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding] UTF8String];
+        }
+    }
+    else
+    {
+        if ([[request.content.userInfo valueForKey: @"data"] isKindOfClass: [NSNumber class]])
+        {
+            NSNumber *value = (NSNumber*)[request.content.userInfo valueForKey: @"data"];
+
+            if (CFBooleanGetTypeID() == CFGetTypeID((__bridge CFTypeRef)(value)))
+            {
+                notificationData->data = (value.intValue == 1) ? "true" : "false";
+            }
+            else
+            {
+                notificationData->data = (char*)[[value description] UTF8String];
+            }
+        }
+        else
+        {
+            notificationData->data = (char*)[[[request.content.userInfo objectForKey: @"data"]description] UTF8String];
+        }
+    }
+
+    return notificationData;
```

Comment:
Are you talking about the `title, subtitle, body etc.` on the `iOSNotificationData`? According to https://developer.apple.com/documentation/foundation/nsstring/1411189-utf8string?language=objc, they should be valid if the `UNNotificationRequest` is valid. 

After checking the code, we pass the `iOSNotificationData` to the registered callbacks while `UNNotificationRequest` is alive. So theoretically it should be fine.

But it might be a problem for the `notification->data` you pointed out above... As it's pointing to a temporary NSString...

Reply:
Fixed at https://github.com/Unity-Technologies/com.unity.mobile.notifications/pull/46/commits/f168e836e4843175c90724834bcdb49b4684d4ed.